docs/admin/reference/settings.md
# Settings

There is one required setting, although specific plugins may have additional required settings.

- `SECRET_KEY <secret-key-setting>`

Pulp uses four types of settings:

- `Django settings` Pulp is configuring
- `Pulp defined settings`
- `Redis settings` Pulp is using
- `Kafka settings`

!!! note
    For more information on how to specify settings see the
    [`Applying Settings`](site:pulpcore/docs/admin/guides/configure-pulp/).


## Django Settings

Below is a list of the most common Django settings Pulp users typically use.
Pulp is a Django project, so any [Django setting] can be set.

### AUTHENTICATION\_BACKENDS

By default, Pulp has two types of authentication enabled, and they fall back for each other:

1. Basic Auth which is checked against an internal users database.
1. Webserver authentication that relies on the webserver to perform the authentication.

To change the authentication types Pulp will use, modify the `AUTHENTICATION_BACKENDS` settings.
See the [Django authentication documentation] for more information.

### DATABASES

By default, Pulp uses PostgreSQL on localhost.
PostgreSQL is the only supported database.
For instructions on how to configure the database, refer to [Django database settings].

### DEFAULT\_FILE\_STORAGE

!!! warning "Deprecated in `pulpcore 3.70`"

    Learn how to update it to [`STORAGES`](#storages) in the [Storages Upgrade Guide].
    Note this must be done between `3.70` and `3.85`.

!!! warning "Removed in `pulpcore 3.85`"

[storages upgrade guide]: https://discourse.pulpproject.org/t/action-required-upgrade-your-storage-settings-before-pulpcore-3-85/2072>

### LOGGING

By default, Pulp logs at an INFO level to syslog.
For all possible configurations please refer to [Django documentation on logging].

Enabling DEBUG logging is a common troubleshooting step.
See the [Enabling Debug Logging] documentation for details on how to do that.

### MEDIA\_ROOT

The location where Pulp will store files.
By default, this is `/var/lib/pulp/media`.

This only affects storage location when `STORAGES['default']['BACKEND']` is set to `pulpcore.app.models.storage.FileSystem`.

See the [storage documentation](site:pulpcore/docs/admin/guides/configure-pulp/configure-storages/) for more info.

It should have permissions of:

- mode: 750
- owner: pulp (the account that pulp runs under)
- group: pulp (the group of the account that pulp runs under)
- SELinux context: `system_u:object_r:pulpcore_var_lib_t:s0`

### SECRET\_KEY

In order to get a pulp server up and running a [Django secret key] *must* be provided.
The following code snippet can be used to generate a random `SECRET_KEY`.

```python linenums="1"
import secrets

chars = 'abcdefghijklmnopqrstuvwxyz0123456789!@#$%^&*(-_=+)'
print(''.join(secrets.choice(chars) for i in range(50)))
```

### STORAGES

!!! note "Added in `3.70`"
    Starting in `3.70`, this should be used in the place of [`DEFAULT_FILE_STORAGE`](#default_file_storage).

Pulp uses [django-storages] to support multiple storage backends.
If no backend is configured, Pulp will by default use the local filesystem (`pulpcore.app.models.storage.FileSystem`).

The storage setting has the form:

```python
STORAGES = {
    "default": {
        "BACKEND": "{storage-class}",
        "OPTIONS": {
            "{option-key}": "{option-value}",
            ...
        },
    },
}
```

To use another backend storage, you'll need to:

1. Set up your storage and gather required credentials and specific configs.
1. Ensure the `django-storage[s3|google|azure]` python package is installed.
This depends on the installation method.
1. Configure the default `BACKEND` storage class.
1. Configure available `OPTIONS` for that backend.

Learn more on the
[Configure Storages](site:pulpcore/docs/admin/guides/configure-pulp/configure-storages/) guide.

Overview of the integration with Pulp:

- **Supported**: We support (test) Amazon S3 and Azure.
- **Untested**: Other backends provided by `django-storages` may work as well, but we provide no guarantee.
- **Known caveat**: Using SFTP Storage is not recommended in Pulp's current state, and doing so can lead to file corruption.
This is because Pulp currently uses coroutines that seem to be incompatible with Django's SFTPStorage implementation.


## Pulp Settings

Pulp defines the following settings itself:

### ALLOWED\_CONTENT\_CHECKSUMS

The list of content-checksums this pulp-instance is **allowed to use**.
By default, the following are used:

```
ALLOWED_CONTENT_CHECKSUMS = ["sha224", "sha256", "sha384", "sha512"]
```

The entire set of supported checksums are:
`md5`, `sha1`, `sha224`, `sha256`, `sha384`, and `sha512`.

!!! warning
    Due to its use as the primary content-identifier, "sha256" **IS REQUIRED**.
    Pulp will fail to start if `"sha256"` is not found in this set.

Pulp can prohibit or allow checksums by setting the `ALLOWED_CONTENT_CHECKSUMS` setting.
Changing this setting requires a few steps.

First, before you change the setting, see how your Pulp instance will be impacted by this change by running:

`pulpcore-manager handle-artifact-checksums --report --checksums sha256,sha512`

Adjust `--checksums` as comma separated list of checksums types to match your needs.

!!! note
    If you already changed `ALLOWED_CONTENT_CHECKSUMS` in pulp settings you can leave out `--checksums`,
    and the checksums will be parsed from Pulp settings.

Before switching, any on-demand repositories containing forbidden checksum digests
need to be synced with `policy=immediate` to populate missing allowed checksums.
This can heavily impact your disk space.
Alternatively, users can remove these offending repository versions followed by orphan cleanup.

If you have artifacts that do not conform to your `ALLOWED_CONTENT_CHECKSUMS` setting, they need to be re-hashed.
You can update them using:

`pulpcore-manager handle-artifact-checksums`

!!! warning
    If Pulp fails to start because forbidden checksums have been identified or required ones are
    missing, run `pulpcore-manager handle-artifact-checksums` command.

### ALLOWED\_EXPORT\_PATHS

One or more real filesystem paths that Exporters can export to.
For example to allow a path of `/mnt/foo/bar/another/folder/` you could specify:

```
ALLOWED_EXPORT_PATHS = ['/mnt/foo/bar']  # only a subpath is needed
```

Defaults to `[]` which means no path is allowed.

### ALLOWED\_IMPORT\_PATHS

One or more real filesystem paths that Remotes with filesystem paths can import from.
For example to allow a remote url of `file:///mnt/foo/bar/another/folder/` you could specify:

```
ALLOWED_IMPORT_PATHS = ['/mnt/foo/bar']  # only a subpath is needed
```

Defaults to `[]`, meaning `file:///` urls are not allowed in any Remote.

### ANALYTICS

If `True`, Pulp will anonymously post analytics information to
[https://analytics.pulpproject.org/](https://analytics.pulpproject.org/) and aids in project decision-making.
See the `analytics docs ` for more info on exactly what is posted along with an example.

Defaults to `True`.

### API\_ROOT

A string containing the prefix for the Pulp API path generated by appending `api/v3/`.
This setting can be used to root the Pulp API in a different namespace on your server.
It must match your reverse proxy configuration and effects hrefs returned from the API.

Defaults to `/pulp/`.
This makes the V3 API by default serve from `/pulp/api/v3/`.

### CACHE\_ENABLED

Store cached responses from the content app into Redis.
This setting improves the performance of the content app under heavy load for similar requests.

Defaults to `False`.

!!! note
    The entire response is not stored in the cache.
    Only the location of the file needed to recreate the response is stored.
    This reduces database queries and allows for many responses to be stored inside the cache.

### CACHE\_SETTINGS

Dictionary with tunable settings for the cache:

- `EXPIRES_TTL` - Number of seconds entries should stay in the cache before expiring.

Defaults to `600` seconds.

!!! note
    Set to `None` to have entries not expire.
    Content app responses are always invalidated when the backing distribution is updated.

### CHUNKED\_UPLOAD\_DIR

A relative path inside the `DEPLOY_ROOT` directory used exclusively for uploaded chunks.
The uploaded chunks are stored in the default storage specified by `STORAGES['default']['BACKEND']`.
This option allows users to customize the actual place where chunked uploads should be stored within the declared storage.
A change to this setting only applies to uploads created after the change.

Defaults to `upload`, which is sufficient for most use cases.

### CONTENT\_APP\_TTL

The number of seconds before a content app should be considered lost.

Defaults to `30` seconds.

### CONTENT\_ORIGIN

A string containing the `protocol`, `fqdn`, and optionally `port` where the content app is reachable by users.
This is used by `pulpcore` and various plugins when referring users to the content app.
For example if the API should redirect users to content at pulp.example.com using HTTPS,
you would set: `https://pulp.example.com` (not including `/pulp/content`, the `CONTENT_PATH_PREFIX`).
Usually this should point to the external address of the reverse proxy.
When set to `None`, the `base_url` for Distributions is a relative path.
This means the API returns relative URLs without the protocol, fqdn, and port.

Default to `None`.

### CONTENT\_PATH\_PREFIX

A string containing the path prefix for the content app.
This is used by the REST API when forming URLs to redirect clients to the content serving app,
and by the content serving application to match incoming URLs.

Defaults to `/pulp/content/`.

### DB\_ENCRYPTION\_KEY

The file location of a symmetric fernet key that Pulp uses to encrypt sensitive fields in the database.
Default location is `/etc/pulp/certs/database_fields.symmetric.key`.

See [Database Encryption] for more details.

### DJANGO\_GUID

Pulp uses `django-guid` to append correlation IDs to logging messages.
For more information on how to configure the `DJANGO_GUID` setting, see the [django-guid settings documentation].
To read more about using correlation id in Pulp, read [our guide](site:pulpcore/docs/user/guides/correlation-id/).

### DOMAIN\_ENABLED

!!! note
    This feature is provided as a tech-preview

Enable the `Domains feature to enable multi-tenancy capabilities <domains>`.
All installed plugins must be Domain compatible for Pulp to start.

Defaults to `False`.

### ENABLED\_PLUGINS

An optional list of plugin names.
If provided, Pulp will limit loading plugins to this list.
If omitted, Pulp will load all installed plugins.

### HIDE\_GUARDED\_DISTRIBUTIONS

If `True`, the distributions, that are protected by a content guard,
will not be shown on the directory listing in the content app.

Defaults to `False`.

### ORPHAN\_PROTECTION\_TIME

The time, specified in minutes, for how long Pulp will hold orphan Content and Artifacts
before they become candidates for deletion by an orphan cleanup task.
This should ideally be longer than your longest running task.
Otherwise any content created during that task could be cleaned up before the task finishes.
Any value different than 0 instanciates a recurring task schedule for this.

Defaults to 1440 minutes (24 hours).

### OTEL\_ENABLED

Toggles the activation of OpenTelemetry instrumentation for monitoring and tracing the application's performance.

Defaults to `False`.

pulpcore/app/settings.py
"""
Django settings for the Pulp Platform application

Never import this module directly, instead `from django.conf import settings`, see
<https://docs.djangoproject.com/en/3.2/topics/settings/#using-settings-in-python-code>

For the full list of settings and their values, see
<https://docs.djangoproject.com/en/3.2/ref/settings/>
"""

import sys

from contextlib import suppress
from importlib import import_module
from importlib.metadata import entry_points
from logging import getLogger
from pathlib import Path

from cryptography.fernet import Fernet
from django.core.exceptions import ImproperlyConfigured
from django.db import connection
from dynaconf import DjangoDynaconf, Dynaconf, Validator

from pulpcore import constants


try:
    import sentry_sdk

    sentry_sdk.init()
except ImportError:
    pass


# Load settings first pass before applying all the defaults to get a grip on ENABLED_PLUGINS.
enabled_plugins_validator = Validator(
    "ENABLED_PLUGINS",
    is_type_of=list,
    len_min=1,
    when=Validator("ENABLED_PLUGINS", must_exist=True),
)

preload_settings = Dynaconf(
    ENVVAR_PREFIX_FOR_DYNACONF="PULP",
    ENV_SWITCHER_FOR_DYNACONF="PULP_ENV",
    ENVVAR_FOR_DYNACONF="PULP_SETTINGS",
    load_dotenv=False,
    validators=[
        enabled_plugins_validator,
    ],
)

# Build paths inside the project like this: BASE_DIR / ...
BASE_DIR = Path(__file__).absolute().parent

WORKER_NAME_TEMPLATE = "{pid}@{hostname}"

# Quick-start development settings - unsuitable for production
# See https://docs.djangoproject.com/en/3.2/howto/deployment/checklist/

# SECURITY WARNING: don't run with debug turned on in production!
DEBUG = False

ALLOWED_HOSTS = ["*"]

DEPLOY_ROOT = Path("/var/lib/pulp")
MEDIA_ROOT = str(DEPLOY_ROOT / "media")  # Django 3.1 adds support for pathlib.Path

# Static files (CSS, JavaScript, Images)
# https://docs.djangoproject.com/en/3.2/howto/static-files/

STATIC_URL = "/assets/"
STATIC_ROOT = DEPLOY_ROOT / STATIC_URL.strip("/")

STORAGES = {
    "default": {
        "BACKEND": "pulpcore.app.models.storage.FileSystem",
    },
    "staticfiles": {
        "BACKEND": "django.contrib.staticfiles.storage.StaticFilesStorage",
    },
}

REDIRECT_TO_OBJECT_STORAGE = True

WORKING_DIRECTORY = DEPLOY_ROOT / "tmp"
FILE_UPLOAD_TEMP_DIR = WORKING_DIRECTORY

CHUNKED_UPLOAD_DIR = "upload"

# List of upload handler classes to be applied in order.
FILE_UPLOAD_HANDLERS = ("pulpcore.app.files.HashingFileUploadHandler",)

# SECURITY WARNING: this should be set to a unique, unpredictable value
SECRET_KEY = "SECRET"

# Key used to encrypt fields in the database
DB_ENCRYPTION_KEY = "/etc/pulp/certs/database_fields.symmetric.key"

# API Root
API_ROOT = "/pulp/"
API_ROOT_REWRITE_HEADER = None

# Application definition

INSTALLED_APPS = [
    # django stuff
    "django.contrib.auth",
    "django.contrib.contenttypes",
    "django.contrib.sessions",
    "django.contrib.messages",
    "django.contrib.staticfiles",
    "django.contrib.postgres",
    "import_export",
    # third-party
    "django_filters",
    "django_guid",
    "drf_spectacular",
    "rest_framework",
    # pulp core app
    "pulpcore.app",
]

# Optional apps that help with development, or augment Pulp in some non-critical way
OPTIONAL_APPS = [
    "crispy_forms",
    "django_extensions",
    "storages",
]

for app in OPTIONAL_APPS:
    # only import if app is installed
    with suppress(ImportError):
        import_module(app)
        INSTALLED_APPS.append(app)

# Select enabled plugins and prepare to load their settings.
enabled_plugins = preload_settings.get("ENABLED_PLUGINS", None)
plugin_settings = []
for entry_point in entry_points(group="pulpcore.plugin"):
    if enabled_plugins is None or entry_point.name in enabled_plugins:
        plugin_app = entry_point.load()
        plugin_settings.append(f"{entry_point.module}.app.settings")
        INSTALLED_APPS += [plugin_app]

MIDDLEWARE = [
    "django_guid.middleware.guid_middleware",
    "django.middleware.security.SecurityMiddleware",
    "whitenoise.middleware.WhiteNoiseMiddleware",
    "django.contrib.sessions.middleware.SessionMiddleware",
    "django.middleware.common.CommonMiddleware",
    "django.middleware.csrf.CsrfViewMiddleware",
    "django.contrib.auth.middleware.AuthenticationMiddleware",
    "django.contrib.messages.middleware.MessageMiddleware",
    "django.middleware.clickjacking.XFrameOptionsMiddleware",
    "pulpcore.middleware.DomainMiddleware",
    "pulpcore.middleware.APIRootRewriteMiddleware",
    "pulpcore.middleware.TaskProfilerMiddleware",
]

AUTHENTICATION_BACKENDS = [
    "django.contrib.auth.backends.ModelBackend",
    "pulpcore.backends.ObjectRolePermissionBackend",
]

ROOT_URLCONF = "pulpcore.app.urls"

TEMPLATES = [
    {
        "BACKEND": "django.template.backends.django.DjangoTemplates",
        "DIRS": [BASE_DIR / "templates"],
        "APP_DIRS": True,
        "OPTIONS": {
            "context_processors": [
                "django.template.context_processors.debug",
                "django.template.context_processors.request",
                "django.contrib.auth.context_processors.auth",
                "django.contrib.messages.context_processors.messages",
            ],
        },
    },
]

WSGI_APPLICATION = "pulpcore.app.wsgi.application"

REST_FRAMEWORK = {
    "URL_FIELD_NAME": "pulp_href",
    "DEFAULT_FILTER_BACKENDS": ("pulpcore.filters.PulpFilterBackend",),
    "DEFAULT_PAGINATION_CLASS": "rest_framework.pagination.LimitOffsetPagination",
    "PAGE_SIZE": 100,
    "DEFAULT_PERMISSION_CLASSES": ("pulpcore.app.access_policy.AccessPolicyFromDB",),
    "DEFAULT_AUTHENTICATION_CLASSES": (
        "pulpcore.app.authentication.BasicAuthentication",
        "rest_framework.authentication.SessionAuthentication",
    ),
    "UPLOADED_FILES_USE_URL": False,
    "DEFAULT_VERSIONING_CLASS": "rest_framework.versioning.URLPathVersioning",
    "DEFAULT_SCHEMA_CLASS": "pulpcore.openapi.PulpAutoSchema",
}

# Password validation
# https://docs.djangoproject.com/en/3.2/ref/settings/#auth-password-validators

AUTH_PASSWORD_VALIDATORS = [
    {"NAME": "django.contrib.auth.password_validation.UserAttributeSimilarityValidator"},
    {"NAME": "django.contrib.auth.password_validation.MinimumLengthValidator"},
    {"NAME": "django.contrib.auth.password_validation.CommonPasswordValidator"},
    {"NAME": "django.contrib.auth.password_validation.NumericPasswordValidator"},
]


# Internationalization
# https://docs.djangoproject.com/en/3.2/topics/i18n/

LANGUAGE_CODE = "en-us"

TIME_ZONE = "UTC"

USE_I18N = "USE_I18N", True

USE_TZ = True


# A set of default settings to use if the configuration file in
# /etc/pulp/ is missing or if it does not have values for every setting

# https://docs.djangoproject.com/en/3.2/ref/settings/#databases
DATABASES = {
    "default": {
        "ENGINE": "django.db.backends.postgresql",
        "NAME": "pulp",
        "USER": "pulp",
        "CONN_MAX_AGE": 0,
    },
}

# Redis default config
REDIS_URL = None
REDIS_HOST = None
REDIS_PORT = None
REDIS_DB = 0
REDIS_PASSWORD = None
REDIS_SSL = False
REDIS_SSL_CA_CERTS = None

# Social_django settings
SOCIAL_AUTH_URL_NAMESPACE = "social"
SOCIAL_AUTH_POSTGRES_JSONFIELD = True
SOCIAL_AUTH_JSONFIELD_CUSTOM = "django.db.models.JSONField"

# https://docs.djangoproject.com/en/3.2/ref/settings/#logging and
# https://docs.python.org/3/library/logging.config.html
LOGGING = {
    "version": 1,
    "disable_existing_loggers": False,
    "formatters": {
        "simple": {"format": "pulp [%(correlation_id)s]: %(name)s:%(levelname)s: %(message)s"}
    },
    "filters": {"correlation_id": {"()": "django_guid.log_filters.CorrelationId"}},
    "handlers": {
        "console": {
            "class": "logging.StreamHandler",
            "formatter": "simple",
            "filters": ["correlation_id"],
        }
    },
    "loggers": {
        "": {
            # The root logger
            "handlers": ["console"],
            "level": "INFO",
            "filters": ["correlation_id"],
        },
        "django_guid": {
            "handlers": ["console"],
            "level": "WARNING",
            "propagate": False,
        },
    },
}

# Custom access policies indexed by the url_pattern of the view.
# This will only take effect when combined with the
# `pulpcore.app.access_policy.AccessPoliciesFromSettings` permission class.
ACCESS_POLICIES = {}

DRF_ACCESS_POLICY = {"reusable_conditions": ["pulpcore.app.global_access_conditions"]}

CONTENT_ORIGIN = None
CONTENT_PATH_PREFIX = "/pulp/content/"

API_APP_TTL = 120  # The heartbeat is called from gunicorn notify (defaulting to 45 sec).
CONTENT_APP_TTL = 30
WORKER_TTL = 30

# Worker implementation type
# Options: "pulpcore" (default, PostgreSQL advisory locks) or "redis" (Redis distributed locks)
WORKER_TYPE = "pulpcore"

# Seconds for a task to finish on semi graceful worker shutdown (approx)
# On SIGHUP, SIGTERM the currently running task will be awaited forever.
# On SIGINT, this value represents the time before the worker will attempt to kill the subprocess.
TASK_GRACE_INTERVAL = 600

# how long to protect ephemeral items in minutes
ORPHAN_PROTECTION_TIME = 24 * 60

# Custom cleanup intervals
# for the following, if set to 0, the corresponding cleanup task is disabled
UPLOAD_PROTECTION_TIME = 0
TASK_PROTECTION_TIME = 0
TMPFILE_PROTECTION_TIME = 0

REMOTE_USER_ENVIRON_NAME = "REMOTE_USER"
REMOTE_USER_OPENAPI_SECURITY_SCHEME = {"type": "mutualTLS"}

AUTHENTICATION_JSON_HEADER = ""
AUTHENTICATION_JSON_HEADER_JQ_FILTER = ""
AUTHENTICATION_JSON_HEADER_OPENAPI_SECURITY_SCHEME = {}

ALLOWED_IMPORT_PATHS = []

ALLOWED_EXPORT_PATHS = []

# https://pulpproject.org/pulpcore/docs/admin/reference/settings/?h=settings#cache_enabled
CACHE_ENABLED = False
CACHE_SETTINGS = {
    "EXPIRES_TTL": 600,  # 10 minutes
}

# The time in seconds a RemoteArtifact will be ignored after failure.
REMOTE_CONTENT_FETCH_FAILURE_COOLDOWN = 5 * 60  # 5 minutes

SPECTACULAR_SETTINGS = {
    "OAS_VERSION": "3.1.1",
    "SERVE_URLCONF": ROOT_URLCONF,
    "DEFAULT_GENERATOR_CLASS": "pulpcore.openapi.PulpSchemaGenerator",
    "DEFAULT_SCHEMA_CLASS": "pulpcore.openapi.PulpAutoSchema",
    "ENUM_ADD_EXPLICIT_BLANK_NULL_CHOICE": False,
    "COMPONENT_SPLIT_REQUEST": True,
    "COMPONENT_NO_READ_ONLY_REQUIRED": True,
    "GENERIC_ADDITIONAL_PROPERTIES": None,
    "DISABLE_ERRORS_AND_WARNINGS": not DEBUG,
    "POSTPROCESSING_HOOKS": [
        "drf_spectacular.hooks.postprocess_schema_enums",
        "pulpcore.openapi.hooks.add_info_hook",
    ],
    "TITLE": "Pulp 3 API",
    "DESCRIPTION": "Fetch, Upload, Organize, and Distribute Software Packages",
    "VERSION": "v3",
    "CONTACT": {
        "name": "Pulp Team",
        "email": "pulp-list@redhat.com",
        "url": "https://pulpproject.org",
    },
    "LICENSE": {
        "name": "GPLv2+",
        "url": "https://raw.githubusercontent.com/pulp/pulpcore/master/LICENSE",
    },
}

# What kinds of checksums is this pulp-instance _allowed to use_ ?
# NOTE : "sha256"" IS REQUIRED - Pulp will fail to start if it is not found in this set
# NOTE: specifying checksums that are not listed under ALL_KNOWN_CONTENT_CHECKSUMS will fail
#       at startup
ALLOWED_CONTENT_CHECKSUMS = ["sha224", "sha256", "sha384", "sha512"]

DEFAULT_AUTO_FIELD = "django.db.models.AutoField"

# Possible diagnostics:
# * "memory"
#    Logs the task process RSS every couple of seconds
# * "pyinstrument"
#    Dumps an HTML profile report produced by pyinstrument, showing time spent in various
#    callstacks. This adds ~10% overhead to the task process and consumes extra memory.
#    Tweaking code might be warranted for some advanced settings.
# * "memray" - Dumps a report produced by memray which logs how much memory was allocated by which
#    lines and functions, at the time of peak RSS of the task process. This adds significant
#    runtime overhead to the task process, 20-40%. Tweaking code might be warranted for
#    some advanced settings.
# NOTE: "memray" and "pyinstrument" require additional packages to be installed on the system.
TASK_DIAGNOSTICS = []  # ["memory", "pyinstrument", "memray"]

ANALYTICS = True

HIDE_GUARDED_DISTRIBUTIONS = False

DOMAIN_ENABLED = False

MAX_CONCURRENT_CONTENT = 25

SHELL_PLUS_IMPORTS = [
    "from pulpcore.app.util import get_domain, get_domain_pk, set_domain, get_url, extract_pk",
    "from pulpcore.tasking.tasks import dispatch, cancel_task, wakeup_worker",
]

# What percentage of available-workers will pulpimport use at a time, max
# By default, use all available workers.
IMPORT_WORKERS_PERCENT = 100

# Kafka settings
KAFKA_BOOTSTRAP_SERVERS = None  # kafka integration disabled by default
KAFKA_TASKS_STATUS_TOPIC = "pulpcore.tasking.status"
KAFKA_TASKS_STATUS_PRODUCER_SYNC_ENABLED = False
KAFKA_PRODUCER_POLL_TIMEOUT = 0.1
KAFKA_SECURITY_PROTOCOL = "plaintext"
KAFKA_SSL_CA_PEM = None
KAFKA_SASL_MECHANISM = None
KAFKA_SASL_USERNAME = None
KAFKA_SASL_PASSWORD = None

# opentelemetry settings
OTEL_ENABLED = False
OTEL_PULP_API_HISTOGRAM_BUCKETS = []

# VulnerabilityReport settings
VULN_REPORT_TASK_LIMITER = 10

# Replaces asyncio event loop with uvloop
UVLOOP_ENABLED = False

# HERE STARTS DYNACONF EXTENSION LOAD (Keep at the very bottom of settings.py)
# Read more at https://www.dynaconf.com/django/

# Validators

storage_validator = (
    Validator("REDIRECT_TO_OBJECT_STORAGE", eq=False)
    | Validator("STORAGES.default.BACKEND", eq="pulpcore.app.models.storage.FileSystem")
    | Validator("STORAGES.default.BACKEND", eq="storages.backends.azure_storage.AzureStorage")
    | Validator("STORAGES.default.BACKEND", eq="storages.backends.s3.S3Storage")
    | Validator("STORAGES.default.BACKEND", eq="storages.backends.s3boto3.S3Boto3Storage")
    | Validator("STORAGES.default.BACKEND", eq="storages.backends.gcloud.GoogleCloudStorage")
)
storage_validator.messages["combined"] = (
    "'REDIRECT_TO_OBJECT_STORAGE=True' is only supported with the local file, S3, GCP or Azure "
    "storage backend configured in STORAGES['default']['BACKEND']."
)

cache_enabled_validator = Validator("CACHE_ENABLED", eq=True)
redis_url_validator = Validator("REDIS_URL", must_exist=True, when=cache_enabled_validator)
redis_host_validator = Validator("REDIS_HOST", must_exist=True, when=cache_enabled_validator)
redis_port_validator = Validator("REDIS_PORT", must_exist=True, when=cache_enabled_validator)
cache_validator = redis_url_validator | (redis_host_validator & redis_port_validator)
cache_validator.messages["combined"] = (
    "CACHE_ENABLED is enabled but it requires to have REDIS configured. Please check "
    "https://pulpproject.org/pulpcore/docs/admin/reference/settings/?h=settings#redis-settings "
    "for more information."
)

sha256_validator = Validator(
    "ALLOWED_CONTENT_CHECKSUMS",
    cont="sha256",
    messages={
        "operations": "ALLOWED_CONTENT_CHECKSUMS MUST contain 'sha256' - Pulp's "
        "content addressable storage relies on sha256 to identify entities."
    },
)

unknown_algs_validator = Validator(
    "ALLOWED_CONTENT_CHECKSUMS",
    condition=lambda x: len(set(x).difference(constants.ALL_KNOWN_CONTENT_CHECKSUMS)) == 0,
    messages={
        "condition": (
            "ALLOWED_CONTENT_CHECKSUMS may only contain algorithms known to pulp - see "
            "constants.ALL_KNOWN_CONTENT_CHECKSUMS for the allowed list."
        )
    },
)

api_root_validator = Validator(
    "API_ROOT",
    condition=lambda x: x.startswith("/") and x.endswith("/"),
    messages={
        "condition": ("The API_ROOT must start and end with a '/', currently it is '{value}'")
    },
)

json_header_auth_class_restframework_validator = Validator(
    "REST_FRAMEWORK__DEFAULT_AUTHENTICATION_CLASSES",
    cont="pulpcore.app.authentication.JSONHeaderRemoteAuthentication",
)

authentication_json_header_validator = Validator(
    "AUTHENTICATION_JSON_HEADER",
    startswith="HTTP_",
    must_exist=True,
    when=json_header_auth_class_restframework_validator,
    messages={"startswith": 'The AUTHENTICATION_JSON_HEADER must start with "HTTP_"'},
)

authentication_json_header_jq_filter_validator = Validator(
    "AUTHENTICATION_JSON_HEADER_JQ_FILTER",
    startswith=".",
    must_exist=True,
    when=json_header_auth_class_restframework_validator,
    messages={"startswith": 'The AUTHENTICATION_JSON_HEADER_JQ_FILTER must start with "."'},
)

json_header_auth_validator = (
    authentication_json_header_validator & authentication_json_header_jq_filter_validator
)

authentication_json_header_openapi_security_scheme_setting_validator = Validator(
    "AUTHENTICATION_JSON_HEADER_OPENAPI_SECURITY_SCHEME", len_min=1
)
authentication_json_header_openapi_security_scheme_validator = Validator(
    "AUTHENTICATION_JSON_HEADER_OPENAPI_SECURITY_SCHEME",
    when=authentication_json_header_openapi_security_scheme_setting_validator,
    is_type_of=dict,
    messages={"is_type_of": "{name} must be a dictionary."},
)

otel_pulp_api_histogram_buckets_validator = Validator(
    "OTEL_PULP_API_HISTOGRAM_BUCKETS",
    is_type_of=list,
    condition=lambda v: all([isinstance(value, float) for value in v]),
    messages={
        "is_type_of": "{name} must be a list.",
        "condition": "All buckets must be declared as a float value",
    },
)


def otel_middleware_hook(settings):
    data = {"dynaconf_merge": True}
    if settings.OTEL_ENABLED:
        data["MIDDLEWARE"] = ["pulpcore.middleware.DjangoMetricsMiddleware"]
    return data


del preload_settings

settings = DjangoDynaconf(
    __name__,
    ENVVAR_PREFIX_FOR_DYNACONF="PULP",
    ENV_SWITCHER_FOR_DYNACONF="PULP_ENV",
    ENVVAR_FOR_DYNACONF="PULP_SETTINGS",
    # Ensure the plugin defaults are loaded before user configs.
    PRELOAD_FOR_DYNACONF=plugin_settings,
    load_dotenv=False,
    validators=[
        api_root_validator,
        cache_validator,
        sha256_validator,
        storage_validator,
        unknown_algs_validator,
        json_header_auth_validator,
        authentication_json_header_openapi_security_scheme_validator,
        otel_pulp_api_histogram_buckets_validator,
    ],
    post_hooks=(otel_middleware_hook,),
)

_logger = getLogger(__name__)


if not (
    Path(sys.argv[0]).name in ["pytest", "sphinx-build"]
    or (len(sys.argv) >= 2 and sys.argv[1] in ["collectstatic", "openapi"])
):
    try:
        with open(DB_ENCRYPTION_KEY, "rb") as key_file:
            Fernet(key_file.read())
    except Exception as ex:
        raise ImproperlyConfigured(
            ("Could not load DB_ENCRYPTION_KEY file '{file}': {err}").format(
                file=DB_ENCRYPTION_KEY, err=ex
            )
        )


FORBIDDEN_CHECKSUMS = set(constants.ALL_KNOWN_CONTENT_CHECKSUMS).difference(
    ALLOWED_CONTENT_CHECKSUMS
)

_SKIPPED_COMMANDS_FOR_CONTENT_CHECKS = [
    "handle-artifact-checksums",
    "migrate",
    "collectstatic",
    "openapi",
]

if not (len(sys.argv) >= 2 and sys.argv[1] in _SKIPPED_COMMANDS_FOR_CONTENT_CHECKS):
    try:
        with connection.cursor() as cursor:
            for checksum in ALLOWED_CONTENT_CHECKSUMS:
                # can't import Artifact here so use a direct db connection
                cursor.execute(
                    f"SELECT count(pulp_id) FROM core_artifact WHERE {checksum} IS NULL LIMIT 1"
                )
                row = cursor.fetchone()
                if row[0] > 0:
                    raise ImproperlyConfigured(
                        (
                            "There have been identified artifacts missing checksum '{}'. "
                            "Run 'pulpcore-manager handle-artifact-checksums' first to populate "
                            "missing artifact checksums."
                        ).format(checksum)
                    )
            for checksum in FORBIDDEN_CHECKSUMS:
                # can't import Artifact here so use a direct db connection
                cursor.execute(
                    f"SELECT count(pulp_id) FROM core_artifact WHERE {checksum} IS NOT NULL LIMIT 1"
                )
                row = cursor.fetchone()
                if row[0] > 0:
                    raise ImproperlyConfigured(
                        (
                            "There have been identified artifacts with forbidden checksum '{}'. "
                            "Run 'pulpcore-manager handle-artifact-checksums' first to unset "
                            "forbidden checksums."
                        ).format(checksum)
                    )

            # warn if there are remote artifacts with checksums but no allowed checksums
            cond = " AND ".join([f"{c} IS NULL" for c in constants.ALL_KNOWN_CONTENT_CHECKSUMS])
            no_checksum_query = f"SELECT pulp_id FROM core_remoteartifact WHERE {cond}"
            cond = " AND ".join([f"{c} IS NULL" for c in ALLOWED_CONTENT_CHECKSUMS])
            cursor.execute(
                f"SELECT count(pulp_id) FROM core_remoteartifact WHERE {cond} AND "
                f"pulp_id NOT IN ({no_checksum_query}) LIMIT 1"
            )
            row = cursor.fetchone()
            if row[0] > 0:
                _logger.warning(
                    (
                        "Warning: detected remote content without allowed checksums. "
                        "Run 'pulpcore-manager handle-artifact-checksums --report' to "
                        "view this content."
                    )
                )

    except ImproperlyConfigured as e:
        raise e
    except Exception:
        # our check could fail if the table hasn't been created yet or we can't get a db connection
        pass
    finally:
        connection.close()

if settings.API_ROOT_REWRITE_HEADER:
    api_root = "/<path:api_root>/"
else:
    api_root = settings.API_ROOT
settings.set("V3_API_ROOT", api_root + "api/v3/")  # Not user configurable
settings.set("V3_DOMAIN_API_ROOT", api_root + "<slug:pulp_domain>/api/v3/")
settings.set("V3_API_ROOT_NO_FRONT_SLASH", settings.V3_API_ROOT.lstrip("/"))
settings.set("V3_DOMAIN_API_ROOT_NO_FRONT_SLASH", settings.V3_DOMAIN_API_ROOT.lstrip("/"))
